### PR TITLE
[WIP] Exercises

### PR DESCRIPTION
**Pull request by JoeriHermans**
Branch for exercises of the course. Both documents (exercises and solutions) will be published in .tex format.

Joeri

### CODE REVIEW

**Comment by JoeriHermans** [comment]
Ping @glouppe 

First session is quite complete I think. Should I add more exercises?

File: exercises/exercise_session_1.tex [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+At this point, we established a method to ensure \emph{reliable} communication. As a result, we are able to coordinate the entities to ensure the correct behaviour of the distributed system. Sadly, reliable \emph{point-to-point} communication is probably not sufficient in a distributed system since more high level features are probably required. For example, let us assume processes $p$, and $q$, where $p$ needs to send message $m$ to $q$. Of course, $p$ could send the message directly to $q$ using our reliable point-to-point abstraction. However, this requires a \emph{direct} connection from $p$ to $q$. It is trivial to see that this the point-to-point protocol assumes some abstraction, i.e., a direct connection between $p$ and $q$ is available. As we will see in later courses, distributed systems, some of which you may or may not use (e.g., BitTorrent), usually place additional layers of abstraction on top of existing infrastructure (albeit with the same functionality) to achieve a specific goal, while at the same time, remove the abstraction when something efficient needs to be done.\\
```
Comment (glouppe) [inline exercises/exercise_session_1.tex]: typo: "that this the"

File: exercises/exercise_session_1.tex [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
```
Comment (glouppe) [inline exercises/exercise_session_1.tex]: You should maybe also comment that processes may crash, in addition to the communication systems. Still, we want a system that remains reliable even if some of nodes crash.

File: exercises/exercise_session_1.tex [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+\item A consumer will eventually consume an element from the queue.
+\item Two consumers cannot consume the same head item concurrently.
+\item A producer has to stop producing if the queue is full.
+\item After a producer stops, it will start again if the queue is not full.
+\end{enumerate}
+
+\subsection*{Exercise 3}
+\label{sec:exercise_3}
+
+\emph{What is the problem vector clocks solved compared to logical clocks? How do they operate differently? In what type of timing abstraction are vector clocks employed?}
```
Comment (glouppe) [inline exercises/exercise_session_1.tex]: Other interesting question: what is the main limitation of vector clocks? (it grows with the number of processes)

File: exercises/exercise_session_1.tex [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+\subsection*{Exercise 5}
+\label{sec:exercise_5}
+
+\emph{We would like to process a very large dataset $\mathcal{D}$ (e.g., $> 4$ PB) of meteorological data to obtain the average temperature of a given location in a given time-frame. We define our computation in a acyclic graph structure $\mathcal{G}$ (drawn on the blackboard). This computation graph (this topic is explained in detail in later lectures) will now be submitted to the cluster, where all nodes will participate in the computation. However, a node fails during the computation. Can you come up with several approaches to recover from such a failure? What are the positive and negative aspects of every approach?}
+
+\subsection*{Exercise 6}
+\label{sec:exercise_6}
+
+\emph{Can you come up with a leader-election abstraction which is resilient against Byzantine adversaries, given the fact that the initial leader is correct? Proof informally (tip: Gibbard–Satterthwaite theorem).}
```
Comment (glouppe) [inline exercises/exercise_session_1.tex:54]: About leader election and failure detection, you could also maybe reason with them regarding why they are equivalent.

**Comment by glouppe** [comment]
This is great! Thank you.

Feel free to merge once my small comments or suggestions are integrated. 

(Also, feel free to add more exercises than what you will eventually do with them.)

File: exercises/exercise_session_1.tex (hunk 4 above)
Comment (JoeriHermans) [reply exercises/exercise_session_1.tex:54]: I'll put this one in my notes, thanks.

**Pull request merged**